src/Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -445,11 +445,24 @@ public void StopStake()
 
             while (!this.stakeCancellationTokenSource.Token.IsCancellationRequested)
             {
-                while (!this.connection.ConnectedNodes.Any() || this.initialBlockDownloadState.IsInitialBlockDownload())
+                // Do not start staking until connected to at least one peer.
+                while (!this.connection.ConnectedNodes.Any())
+                {
+                    this.logger.LogTrace("Waiting to be connected with at least one network peer...");
+                    await Task.Delay(TimeSpan.FromMilliseconds(this.minerSleep), this.stakeCancellationTokenSource.Token).ConfigureAwait(false);
+                }
+
+                // Do not start staking if we're in IBD.
+                while (this.initialBlockDownloadState.IsInitialBlockDownload())
                 {
-                    if (!this.connection.ConnectedNodes.Any()) this.logger.LogTrace("Waiting to be connected with at least one network peer...");
-                    else this.logger.LogTrace("Waiting for IBD to complete...");
+                    this.logger.LogTrace("Waiting for IBD to complete...");
+                    await Task.Delay(TimeSpan.FromMilliseconds(this.minerSleep), this.stakeCancellationTokenSource.Token).ConfigureAwait(false);
+                }
 
+                // Wait until we have at least one connected peer who's headers are synced with ours.
+                while (!this.connection.ConnectedNodes.Any(x => x.Behavior<ChainHeadersBehavior>().IsSynced()))
```

Comment:
why is this separated from the first loop?

would it be not enough, in original code, to just add the condition to Any() and change the comment?

these should not be separated loops anyway, what if IBD says true and then false again?

Reply:
Fixed.